File: src/test/java/mission/step3/ProductServiceTest.java
package mission.step3;

import kitchenpos.application.ProductService;
import kitchenpos.domain.*;
import kitchenpos.infra.PurgomalumClient;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
class ProductServiceTest {

    @Mock
    private ProductRepository productRepository;

    @Mock
    private MenuRepository menuRepository;

    @Mock
    private PurgomalumClient purgomalumClient;

    @InjectMocks
    private ProductService productService;

    @Test
    @DisplayName("상품을 생성한다")
    void create() {
        // given
        Product request = new Product();
        request.setName("아메리카노");
        request.setPrice(BigDecimal.valueOf(4000));

        given(purgomalumClient.containsProfanity(anyString()))
                .willReturn(false);
        given(productRepository.save(any(Product.class)))
                .willAnswer(invocation -> {
                    Product saved = invocation.getArgument(0);
                    saved.setId(UUID.randomUUID());
                    return saved;
                });

        // when
        Product created = productService.create(request);

        // then
        assertThat(created.getName()).isEqualTo("아메리카노");
        assertThat(created.getPrice()).isEqualByComparingTo(BigDecimal.valueOf(4000));
        verify(productRepository).save(any(Product.class));
    }

    @Test
    @DisplayName("상품 이름에 부적절한 단어가 포함되어 있으면 예외가 발생한다")
    void createWithProfanity() {
        // given
        Product request = new Product();
        request.setName("부적절한 이름");
        request.setPrice(BigDecimal.valueOf(4000));

        given(purgomalumClient.containsProfanity(anyString()))
                .willReturn(true);

        // when & then
        assertThatThrownBy(() -> productService.create(request))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    @DisplayName("상품 가격이 null이면 예외가 발생한다")
    void createWithNullPrice() {
        // given
        Product request = new Product();
        request.setName("아메리카노");
        request.setPrice(null);

        // when & then
        assertThatThrownBy(() -> productService.create(request))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    @DisplayName("상품 가격이 음수이면 예외가 발생한다")
    void createWithNegativePrice() {
        // given
        Product request = new Product();
        request.setName("아메리카노");
        request.setPrice(BigDecimal.valueOf(-1000));

        // when & then
        assertThatThrownBy(() -> productService.create(request))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    @DisplayName("상품의 가격을 변경한다")
    void changePrice() {
        // given
        UUID productId = UUID.randomUUID();
        Product product = new Product();
        product.setId(productId);
        product.setPrice(BigDecimal.valueOf(4000));

        Product request = new Product();
        request.setPrice(BigDecimal.valueOf(4500));

        given(productRepository.findById(productId))
                .willReturn(Optional.of(product));
        given(menuRepository.findAllByProductId(productId))
                .willReturn(List.of());

        // when
        Product updated = productService.changePrice(productId, request);

        // then
        assertThat(updated.getPrice()).isEqualByComparingTo(BigDecimal.valueOf(4500));
    }

    @Test
    @DisplayName("상품 가격 인상으로 메뉴 가격이 원가보다 낮아지면 판매 중지로 변경된다")
    void changePriceToHigherThanMenuPrice() {
        // given
        UUID productId = UUID.randomUUID();

        Product product = new Product();
        product.setId(productId);
        product.setPrice(BigDecimal.valueOf(4000));

        Menu menu = new Menu();
        menu.setPrice(BigDecimal.valueOf(9000));
        menu.setDisplayed(true);

        MenuProduct menuProduct = new MenuProduct();
        menuProduct.setQuantity(2);
        menuProduct.setProduct(product);
        menu.setMenuProducts(List.of(menuProduct));

        Product request = new Product();
        request.setPrice(BigDecimal.valueOf(5000));

        given(productRepository.findById(productId))
                .willReturn(Optional.of(product));
        given(menuRepository.findAllByProductId(productId))
                .willReturn(List.of(menu));

        // when
        productService.changePrice(productId, request);

        // then
        assertThat(product.getPrice()).isEqualTo(BigDecimal.valueOf(5000));
        assertThat(menu.isDisplayed()).isTrue();
    }

    @Test
    @DisplayName("모든 상품을 조회한다")
    void findAll() {
        // given
        Product product1 = new Product();
        product1.setId(UUID.randomUUID());
        product1.setName("아메리카노");
        product1.setPrice(BigDecimal.valueOf(4000));

        Product product2 = new Product();
        product2.setId(UUID.randomUUID());
        product2.setName("카페라떼");
        product2.setPrice(BigDecimal.valueOf(4500));

        given(productRepository.findAll())
                .willReturn(List.of(product1, product2));

        // when
        List<Product> products = productService.findAll();

        // then
        assertThat(products).hasSize(2);
        assertThat(products.get(0).getName()).isEqualTo("아메리카노");
        assertThat(products.get(1).getName()).isEqualTo("카페라떼");
    }
}

File: src/test/kotlin/kitchenops/application/ProductServiceTest.kt
package kitchenops.application

import io.kotest.matchers.should
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import kitchenpos.application.ProductService
import kitchenpos.domain.Menu
import kitchenpos.domain.MenuProduct
import kitchenpos.domain.MenuRepository
import kitchenpos.domain.Product
import kitchenpos.domain.ProductRepository
import kitchenpos.infra.PurgomalumClient
import spec.BaseUnitSpec
import java.math.BigDecimal
import java.util.*

internal class ProductServiceTest : BaseUnitSpec({

    val productRepository = mockk<ProductRepository>()
    val menuRepository = mockk<MenuRepository>()
    val purgomalumClient = mockk<PurgomalumClient>()

    val sut = ProductService(productRepository, menuRepository, purgomalumClient)

    context("상품을 생성할 수 있다") {
        test("가격을 지정하지 않으면 생성할 수 없다") {
            // given
            val request = Product().apply { price = null }

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("가격이 0원 미만이면 생성할 수 없다") {
            // given
            val request = Product().apply { price = BigDecimal("-1") }

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("이름을 지정하지 않으면 생성할 수 없다") {
            // given
            val request = Product().apply {
                price = BigDecimal("100")
                name = null
            }

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        // TODO: empty name에 동작하지 않아야 할 것 같은데 버그인 듯 하다
        test("이름이 비어있으면 생성할 수 없다").config(enabled = false) {
            // given
            val request = Product().apply {
                price = BigDecimal("100")
                name = ""
            }

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("상품의 이름에 욕설이 포함되면 생성할 수 없다") {
            // given
            val request = Product().apply {
                price = BigDecimal("10")
                name = "hell"
            }

            every { purgomalumClient.containsProfanity("hell") } returns true

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("생성한다") {
            // given
            val request = Product().apply {
                price = BigDecimal("1000")
                name = "chicken"
            }

            every { purgomalumClient.containsProfanity("chicken") } returns false
            every { productRepository.save(match { it.price == BigDecimal("1000") && it.name == "chicken" }) } returns request

            // when
            val actual = sut.create(request)

            // then
            actual should {
                it.price shouldBe BigDecimal("1000")
                it.name shouldBe "chicken"
            }

            verify(exactly = 1) { productRepository.save(any()) }
        }
    }

    context("상품을 전체 조회할 수 있다") {
        test("전체 조회한다") {
            // given
            val product1 = Product().apply {
                price = BigDecimal("19900")
                name = "fried chicken"
            }
            val product2 = Product().apply {
                price = BigDecimal("3000")
                name = "cheese ball"
            }

            every { productRepository.findAll() } returns listOf(product1, product2)

            // when
            val actual = sut.findAll()

            // then
            actual shouldBe listOf(product1, product2)
        }
    }

    context("상품의 가격을 변경할 수 있다") {
        test("상품의 가격을 지정하지 않으면 변경할 수 없다") {
            // given
            val request = Product().apply { price = null }

            // when
            val actual = runCatching { sut.changePrice(request.id, request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("상품의 가격이 0원 미만이면 변경할 수 없다") {
            // given
            val request = Product().apply { price = BigDecimal("-1") }

            // when
            val actual = runCatching { sut.changePrice(request.id, request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("존재하지 않는 상품은 변경할 수 없다") {
            // given
            val request = Product().apply { price = BigDecimal("100") }

            every { productRepository.findById(request.id) } returns Optional.empty()

            // when
            val actual = runCatching { sut.changePrice(request.id, request) }

            // then
            actual.exceptionOrNull() shouldBe NoSuchElementException()
        }

        test("상품을 사용하는 메뉴의 가격이 메뉴를 구성하는 상품 가격의 총 합보다 클 경우 메뉴를 숨긴다") {
            // given
            val request = Product().apply { price = BigDecimal("500") }
            every { productRepository.findById(request.id) } returns Optional.of(request)

            val product1 = Product().apply {
                id = request.id
                price = request.price
            }
            val menuProduct1 = MenuProduct().apply {
                product = product1
                quantity = 2
            }
            val menu = Menu().apply {
                price = BigDecimal("600")
                menuProducts = listOf(menuProduct1)
            }
            every { menuRepository.findAllByProductId(request.id) } returns listOf(menu)

            // when
            val actual = sut.changePrice(request.id, request)

            // then
            actual.price shouldBe BigDecimal("500")
            menu.isDisplayed shouldBe false
        }

        test("가격을 변경한다") {
            // given
            val request = Product().apply { price = BigDecimal("300") }

            every { productRepository.findById(request.id) } returns Optional.of(request)
            every { menuRepository.findAllByProductId(request.id) } returns emptyList()

            // when
            val actual = sut.changePrice(request.id, request)

            // then
            actual.price shouldBe BigDecimal("300")
        }
    }
})
